File: old java code/src/GaiaNet/GaiaNet/DHTLikeNet.java
package GaiaNet.GaiaNet;

import GaiaNet.NetTransfer.GaiaType;
import GaiaNet.NetTransfer.MultiThreadServer;

import java.io.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.logging.FileHandler;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

public class DHTLikeNet<toString> {
    private static final Logger logNet = Logger.getLogger( DHTLikeNet.class.getName() );
    public FileHandler fileHandler;  // For log file.
    public Node node = null;
    public ArrayList<String> masterNodesIp;
    public ArrayList<String> neighbours;

    public DHTLikeNet(Node node, Config config){
        this.node = node;
        this.masterNodesIp = (ArrayList<String>) config.masterNodesIp;
        this.neighbours = (ArrayList<String>) config.neighbours;

        try {
            File logDir = new File("./log/");
            if (!logDir.exists())   logDir.mkdir();
            this.fileHandler = new FileHandler("./log/Net.log");   // For log file.
            this.fileHandler.setFormatter(new SimpleFormatter());
            this.logNet.addHandler(fileHandler);
        } catch (IOException e) {
            e.printStackTrace();
        }

        logNet.info("net initialize: " + this.toString());
        joinNet();
    }

    public int joinNet(){
        for (String serverIP : this.masterNodesIp) {
            logNet.info("joinNet, send to " + serverIP);
            Socket sock = null;
            try {
                sock = new Socket(serverIP, 9190);
                DataOutputStream dos = new DataOutputStream(sock.getOutputStream());
                dos.writeInt(GaiaType.GaiaNet.ordinal());
                dos.writeInt(NetRequest.JoinNet.ordinal());

                ObjectOutputStream oos = new ObjectOutputStream(sock.getOutputStream());
                ObjectInputStream ois = new ObjectInputStream(sock.getInputStream());
                oos.writeObject(this.node);
                Node remoteNode = (Node) ois.readObject();
                logNet.info("master response: " + remoteNode.toString());

                boolean isMyneigh = isMyNeighbour(remoteNode);
                if (isMyneigh) joinMyNeighbourList(remoteNode);
            } catch (IOException | ClassNotFoundException e) {
                e.printStackTrace();
            }finally {
                try { if (sock != null) sock.close(); }
                catch (IOException e) { e.printStackTrace();}
            }
        }

        return 0; // normal
    }


    public void joinNetResponse(Socket socket, InputStream is, OutputStream os){
        try {
            ObjectOutputStream oos = new ObjectOutputStream(os);
            ObjectInputStream ois = new ObjectInputStream(is);
            Node remoteNode = (Node) ois.readObject();
            oos.writeObject(this.node);
            logNet.info("A node ask for joining net: " + remoteNode.toString());

            boolean isMyneigh = isMyNeighbour(remoteNode);
            if (isMyneigh) joinMyNeighbourList(remoteNode);
        } catch (IOException | ClassNotFoundException e) {
            e.printStackTrace();
        }finally {
            try { if (socket != null) socket.close(); }
            catch (IOException e) { e.printStackTrace();}
        }

    }


    public void Config(){ }

    public void handle(Socket socket, Logger logger){
        try {
            OutputStream os = socket.getOutputStream();
            InputStream is = socket.getInputStream();
            DataInputStream dis = new DataInputStream(is);
            int flag = dis.readInt();
            NetRequest type = NetRequest.values()[flag];

            if (type == NetRequest.JoinNet) joinNetResponse(socket, is, os);
            else if (type == NetRequest.LeaveNet);
            else if (type == NetRequest.NearestNodes);

            socket.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void joinMyNeighbourList(Node node) {
    }

    private boolean isMyNeighbour(Node node) {
        return false;
    }

    private String getMyNearestNodes() {
        return this.neighbours.toString();
    }


    public int distance(Node node1, Node node2){
        return 0;
    }

    public int leaveNet(){
        return 0;
    }

    public void nearest(Node node) {

    }

    @Override
    public String toString(){
        String s = "";
        s += "Node Name in DHTLikeNet: " + this.node.nodeName + "\n";
        s += "masterNodesIp: " + this.masterNodesIp + "\n";
        s += "neighbours: " + this.neighbours.toString() + "\n";
        return s;
    }
}

File: GaiaNets/DHTLikeNet.cs
using System;
using System.IO;
using System.Net;
using System.Linq;
using System.Collections.Generic;
using System.Net.Sockets;
using GaiaNet.BasicNet;

namespace GaiaNet.GaiaNets
{
    public enum NetRequest : byte { JoinNet = 0x01, LeaveNet = 0x02, NearestNodes = 0x03 };
    public class DHTLikeNet
    {
        public static Node node = null;
        public static Dictionary<String, String> masterNodesIp = new Dictionary<String, String>(StringComparer.OrdinalIgnoreCase);
        public static Dictionary<String, String> neighbours = new Dictionary<String, String>(StringComparer.OrdinalIgnoreCase);
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(
            System.Reflection.MethodBase.GetCurrentMethod().ReflectedType);

        public DHTLikeNet(Node node1)
        {
            node = node1;
            masterNodesIp = Config.masterNodesIp;
            neighbours = Config.neighbours;
            log.Info("DHTLikeNet initialize: " + this.ToString());
            Console.WriteLine("DHTLikeNet initialize: " + this.ToString());
            // JoinNet();
        }

        public int JoinNet()
        {
            if (masterNodesIp.Count <= 0) return 1;
            foreach (String serverIP in masterNodesIp.Values)
            {
                log.Info("joinNet, send to " + serverIP);
                Socket sock = null;
                try
                {
                    sock = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                    sock.Connect(serverIP, 9190);
                    sock.Send(new byte[] { (byte)NetType.GaiaNet, (byte)NetRequest.JoinNet });
                    sock.Send(node.ToBytes());
                    Node remoteNode = null; // (Node) sock.Receive();
                    log.Info("master response: " + remoteNode.ToString());

                    Boolean isMyneigh = IsMyNeighbour(remoteNode);
                    if (isMyneigh) JoinMyNeighbourList(remoteNode);
                }
                catch (IOException e)
                {
                    log.Error(e);
                    return 1;
                }
                finally
                {
                    if (sock != null) sock.Close();
                }
            }

            return 0; // normal
        }


        public void JoinNetResponse(Socket socket)
        {
            try
            {
                Node remoteNode = null; //(Node) ois.readObject();
                // oos.writeObject(this.node);
                log.Info("A node ask for joining net: " + remoteNode.ToString());

                Boolean isMyneigh = IsMyNeighbour(remoteNode);
                if (isMyneigh) JoinMyNeighbourList(remoteNode);
            }
            catch (IOException e)
            {
                log.Error(e);
            }
            finally
            {
                if (socket != null) socket.Close();
            }

        }

        public void handle(Socket socket)
        {
            try
            {
                // socket.Receive();
                // int flag = dis.readInt();
                // NetRequest type = NetRequest.values()[flag];

                // if (type == NetRequest.JoinNet) JoinNetResponse(socket, is, os);
                // else if (type == NetRequest.LeaveNet);
                // else if (type == NetRequest.NearestNodes);

                socket.Close();
            }
            catch (IOException e)
            {
                log.Error(e);
            }
        }

        private void JoinMyNeighbourList(Node node)
        {
        }

        private Boolean IsMyNeighbour(Node node)
        {
            return false;
        }

        private String GetMyNearestNodes()
        {
            return neighbours.ToString();
        }


        public int distance(Node node1, Node node2)
        {
            return 0;
        }

        public int leaveNet()
        {
            return 0;
        }

        public void nearest(Node node)
        {

        }

        public override String ToString()
        {
            String s = "\n";
            s += "\t- Node Name: " + node.nodeName + "\n";
            s += "\t- masterNodesIp: " + String.Join(", ", masterNodesIp.Select(x => x.Key + " = " + x.Value).ToArray()) + "\n";
            s += "\t- neighbours: " + String.Join(", ", neighbours.Select(x => x.Key + " = " + x.Value).ToArray()) + "\n";
            return s;
        }


        /// <summary> 
        /// Parse the ip string 192.168.1.1[:8081], if the port is not specified, the default port is "serverPort".
        /// </summary>
        /// <param name="ipstr">The string of ip[:port]</param>
        /// <returns>IPEndPoint (ip and port)</returns>
        public static IPEndPoint ParseIPStr(string ipstr)
        {
            string ip = "127.0.0.1";
            int port = Config.serverPort;
            if (ipstr.Contains(":"))
            {
                string[] ip_port = ipstr.Split(':');
                if (ip_port.Length == 2)
                {
                    ip = ip_port[0];
                    int.TryParse(ip_port[1], out port);
                }
            }
            else { ip = ipstr; }
            IPEndPoint iPEndPoint = new IPEndPoint(IPAddress.Parse(ip), port);
            return iPEndPoint;
        }

        /// <summary> Get the IP from the node name.
        /// </summary>
        /// <param name="nodename"> node name. </param>
        /// <returns>IPEndPoint (ip and port)</returns>
        public static IPEndPoint GetNodeIP(string nodename)
        {
            IPEndPoint iPEndPoint = null;
            if (nodename.Trim().Equals(node.nodeName.Trim(), StringComparison.OrdinalIgnoreCase))
            {
                iPEndPoint = new IPEndPoint(IPAddress.Parse("127.0.0.1"), Config.serverPort);
            }
            else if (DHTLikeNet.masterNodesIp.ContainsKey(nodename))
            {
                iPEndPoint = ParseIPStr(DHTLikeNet.masterNodesIp[nodename]);
            }
            else if (DHTLikeNet.neighbours.ContainsKey(nodename))
            {
                iPEndPoint = ParseIPStr(DHTLikeNet.neighbours[nodename]);
            }
            return iPEndPoint;
        }
    }
}